x/exchange/msg.go

```diff
@@ -0,0 +1,464 @@
+}
+
+func (m MsgMarketUpdateEnabledRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	// Nothing to validate for the AcceptingOrders field.
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateEnabledRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketUpdateUserSettleRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	// Nothing to validate for the AllowUserSettlement field.
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateUserSettleRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketManagePermissionsRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	if m.HasUpdates() {
+		for _, addrStr := range m.RevokeAll {
+			if _, err := sdk.AccAddressFromBech32(addrStr); err != nil {
+				errs = append(errs, fmt.Errorf("invalid revoke-all address %q: %w", addrStr, err))
+			}
+		}
+		if ContainsString(m.RevokeAll, m.Admin) {
+			errs = append(errs, fmt.Errorf("message administrator %s cannot revoke all of their permissions", m.Admin))
+		}
+
+		if err := ValidateAccessGrantsField("to-revoke", m.ToRevoke); err != nil {
+			errs = append(errs, err)
+		}
+
+		toRevokeByAddr := make(map[string]AccessGrant, len(m.ToRevoke))
+		for _, ag := range m.ToRevoke {
+			if ContainsString(m.RevokeAll, ag.Address) {
+				errs = append(errs, fmt.Errorf("address %s appears in both the revoke-all and to-revoke fields", ag.Address))
+			}
+			if ag.Address == m.Admin && ag.Contains(Permission_permissions) {
+				errs = append(errs, fmt.Errorf("message administrator %s cannot revoke their own ability to manage permissions",
+					ag.Address))
+			}
+			toRevokeByAddr[ag.Address] = ag
+		}
+
+		if err := ValidateAccessGrantsField("to-grant", m.ToGrant); err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, ag := range m.ToGrant {
+			toRev, ok := toRevokeByAddr[ag.Address]
+			if ok {
+				for _, perm := range ag.Permissions {
+					if toRev.Contains(perm) {
+						errs = append(errs, fmt.Errorf("address %s has both revoke and grant %q", ag.Address, perm.SimpleString()))
+					}
+				}
+			}
+		}
+	} else {
```

Comment:
You can actually get rid of this else statement and adding an error check before this branch.

```
if !m.HasUpdates() {
  errs = append(errs, errors.New("no updates"))
}
```

Reply:
I thought about that, but wanted only the one return at the end of the function.

Reply:
That makes sense. I was just pointing it out because it goes against the [go convention](https://go.dev/doc/effective_go).

Reply:
Go convention doesn't say that you should use early returns and loop control statements, it just says that if you do, put them first and leave the `else` clause off. In this case, I don't want an early return and there isn't a loop control statement. I only want the one return at the end so that it's clear that everything is validated even if there's an already known problem.

I could put a label before the return and `if !m.HasUpdates() { errs = ...;  goto ...; }`, but I imagine there'd be more complaints about that than this.

Reply:
Whoops my mistake on that, thought I saw reference to "the happy path is left-aligned" in there too. That quote is what I'm referring to though. Are you trying to have a single point of return to aggregate errors? I don't think there are any errors to aggregate in the rainy day path for this function.